### PR TITLE
Change Hash implementation

### LINKED ISSUE #4557: Reimplementation a Hash using open addressing
Crystal Hash(K, V) should be reimplemented using open addressing and powers of two.

To be read: 

* http://www.idryman.org/blog/2017/05/03/writing-a-damn-fast-hash-table-with-tiny-memory-footprints/.
* https://github.com/konovod/testhash
* https://github.com/yxhuvud/crystal-hash
* https://developers.redhat.com/blog/2017/02/27/towards-faster-ruby-hash-tables/
* (sometimes 508) http://blog.redpanthers.co/behind-scenes-hash-table-performance-ruby-2-4

Thanks to @RX14, @konovod, @yxhuvud for links.

Refs #2817, #3932, #2331 and https://github.com/crystal-lang/crystal/pull/4033#issuecomment-307779149 to continue discussion and implementation.

### PR DESCRIPTION
**Pull request by funny-falcon**
use array of entries + open addressing index of entries.

Closes #4557

### CODE REVIEW

File: src/hash.cr [hunk 1]
```diff
@@ -8,19 +8,28 @@ class Hash(K, V)
   include Iterable({K, V})
 
   getter size : Int32
-  @buckets_size : Int32
-  @first : Entry(K, V)?
-  @last : Entry(K, V)?
+  @sz : UInt8
```
Comment (RX14) [inline src/hash.cr]: This is badly named.

File: src/hash.cr [hunk 2]
```diff
@@ -618,17 +599,26 @@ class Hash(K, V)
...
   def first_value
-    @first.not_nil!.value
+    nil.not_nil! if empty?
+    @entries[@first].value
   end
 
   # Similar to `#first_key?`, but returns its value.
   def first_value?
-    @first.try &.value
+    unless empty?
```
Comment (RX14) [inline src/hash.cr]: `return nil if empty?`/`return yield if empty?` is more readable and ditto down the page.

File: src/hash.cr [hunk 3]
```diff
@@ -804,113 +809,256 @@ class Hash(K, V)
...
     self.each do |k, v|
       hash[v] = k
     end
     hash
   end
 
-  protected def find_entry(key)
-    return nil if empty?
+  @[AlwaysInline]
+  private def iter_entries
```
Comment (RX14) [inline src/hash.cr]: `each_entry`?

File: src/hash.cr [hunk 4]
```diff
@@ -804,113 +809,256 @@ class Hash(K, V)
...
+    if new_nindex != old_nindex
+      @index = @index.realloc(new_nindex)
+    end
+    if new_nentries > old_nentries
+      (@entries + old_nentries).clear(new_nentries - old_nentries)
+    end
+    @sz = newsz
+  end
+
+  private def need_shrink(size : Int32, sz : UInt8) : Bool
```
Comment (RX14) [inline src/hash.cr]: `needs_shrink?`?

File: src/hash.cr [hunk 5]
```diff
@@ -804,113 +809,256 @@ class Hash(K, V)
...
-    getter key : K
-    property value : V
+  def hash_key(key)
+    h = key.hash.to_u64
+    (h >> 1) + 1
+  end
+
+  private def nindex(sz)
+    mask = SIZES[sz].indexmask
+    mask + (mask != 0 ? 1 : 0)
```
Comment (RX14) [inline src/hash.cr]: `mask == 0 ? 0 : 1`

File: src/hash.cr [hunk 6]
```diff
@@ -804,113 +809,256 @@ class Hash(K, V)
...
 
-  private class Entry(K, V)
-    getter key : K
-    property value : V
+  def hash_key(key)
+    h = key.hash.to_u64
+    (h >> 1) + 1
+  end
+
+  private def nindex(sz)
```
Comment (RX14) [inline src/hash.cr]: This isn't a very crystally name.

**Comment by funny-falcon** [comment]
Worse problem: doesn't look it is faster :-(
While some microbenchmarks I've made shows it is a bit faster, http helloworld is slower, and std_spec is also a bit slower.

And 32bit compiler test didn't pass in Travis-CI.

**Comment by funny-falcon** [comment]
Looks like, reallocation costs too much.
Probably could be better if entries are allocated by chunks instead of by huge array.
I'll close this PR now until improving.

**Comment by akzhan** [comment]
I think that Hash may have two different implementation for small (<7) and other sizes.

I'm sure that small hashes have a big value.

**Comment by asterite** [comment]
I benchmarked this PR and it works twice as fast as the current Hash. Here's my benchmark:

```crystal
a = {} of Int32 => String

strings = Array.new(10_000_000, &.to_s)

time = Time.now
strings.each_with_index do |s, i|
  a[i] = s
end
puts Time.now - time
```

Remember to run this with `bin/crystal` in master and `bin/crystal` with your branch, so we are sure both use funny hash.

It would be interesting to see your benchmarks.

I think this Hash should be more efficient because in the current Hash every time you add a key-value pair there's an allocation for the node. In your implementation that doesn't seem to be the case.

**Comment by funny-falcon** [comment]
It is expected for single huge hash to be faster.

But http helloworld ( from crystal-lang.org main page) prouces many small hashes.

**Comment by RX14** [comment]
Testing compiler compile performance (with `--no-codegen`) before and after is likely to be a good indicator of whether this hash is better just in theory or in practice. Microbenchmarks are only so useful.

I don't like the idea of having a rigid cutoff point with 2 different hash implementations. There must be a way to make it fast at any size.

**Comment by funny-falcon** [comment]
I have an idea to use chunked allocaion, ie allocate by chunks of 8 entries.
Probably it will be faster than current hash implementation, and not too slower than this PR on huge hashes. I will investigate it.

**Comment by RX14** [comment]
@funny-falcon Assuming that a larger chunk size would increase performance, you could change chunk size based on the size of the hash to keep the memory overhead bounded when small and performance overhead miniscule when large.

**Comment by akzhan** [comment]
What about implementation of pessimistic downsize politics?

**Comment by funny-falcon** [comment]
@akzhan , what do you mean?

**Comment by akzhan** [comment]
Do relocations on increase, but rarely on decrease.
пт, 10 нояб. 2017 г. в 20:12, Sokolov Yura <notifications@github.com>:

**Comment by funny-falcon** [comment]
That was already in this patch: it resized down only if `size < capacity / 4`

**Comment by funny-falcon** [comment]
Looks like chunked version is not slower than current implementation even on http hello world, so that I reopen pull request.

**Comment by RX14** [comment]
@funny-falcon Please give us some documentation on how this works to work with!

File: src/hash.cr [hunk 7]
```diff
@@ -8,19 +8,26 @@ class Hash(K, V)
...
-    initial_capacity = 11 if initial_capacity < 11
-    initial_capacity = initial_capacity.to_i
-    @buckets = Pointer(Entry(K, V)?).malloc(initial_capacity)
-    @buckets_size = initial_capacity
     @size = 0
+    @format = 0_u8
+    @rebuild_num = 0_u16
+    @first = 0_u32
+    @last = 0_u32
+    @index = Pointer(UInt32).new(0)
```
[the remark this answers is not in the document]
Comment (RX14) [reply src/hash.cr]: You mean `Pointer.null`.

[the remark this answers is not in the document]
Comment (Sija) [reply src/hash.cr]: @RX14 right, my bad.

File: src/hash.cr [hunk 8]
```diff
@@ -804,113 +806,317 @@ class Hash(K, V)
...
+  end
+
+  @[AlwaysInline]
+  private def each_entry
+    return if empty?
+    rnum = @rebuild_num
+    chunks = chunks_ptr
+    @first.upto(@last - 1) do |i|
+      chunk = chunks[i / CHUNK]
+      entry = chunk + i % CHUNK
```
Comment (yxhuvud) [inline src/hash.cr]: I'm not certain there is a noticeable difference or if llvm is smart enough to notice this by itself, but perhaps 
```
chunk_index, offset = i.divmod(CHUNK)
```

Comment (funny-falcon) [reply src/hash.cr]: llvm should optimize division and module by 8 to shifts and mask (ie `i >> 3` and `i & 7`)

File: src/hash.cr [hunk 9]
```diff
@@ -8,19 +8,26 @@ class Hash(K, V)
...
-    @buckets_size = initial_capacity
     @size = 0
+    @format = 0_u8
+    @rebuild_num = 0_u16
+    @first = 0_u32
+    @last = 0_u32
+    @index = Pointer(UInt32).new(0)
+    @entries = Pointer(Void).new(0)
     @block = block
+    unless initial_capacity.nil?
```
[the remark this answers is not in the document]
Comment (akzhan) [reply src/hash.cr]: it may be zero, isn't it?

[the remark this answers is not in the document]
Comment (RX14) [reply src/hash.cr]: only `nil`, `Pointer.null` and `false` are falsey in crystal. `0` is truthy, which means the only practical situation where you need to use `nil?` is `Bool?`, which is fairly rare and a situation often better described with an enum.

**Comment by akzhan** [comment]
Is it GTG?

**Comment by funny-falcon** [comment]
It plays bad with conservative GC on 32 bit platform:
- stored hashsum could be counted as pointers (failed weakptr test on travis ci). That is why in first version I stored hashsums in separate array (GC then knows to not look inside).
- index array is five-ten times greater than bins in current implementation. Looks like GC doesn't like such big arrays.

So, while small scripts will still work, larger applications become less reliable on 32bit platform with this patch.

So, if support of 32 bit platform is important, this patch should be improved. I don't know, how it should be done. I'm still thinking.

**Comment by funny-falcon** [comment]
Doubtfuly Crystal will gain precise GC soon.

**Comment by akzhan** [comment]
Have you seen #5271?

**Comment by pnloyd** [comment]
@akzhan, not totally clear if anyone can or wants to pushed that forward ATM though.

**Comment by RX14** [comment]
I wouldn't focus on the performance of 32-bit too much, but it should work and be stable. 32bit is almost obsolete on x86, but on arm it'll still be a while until 32bit SoCs stop being sold.

**Comment by funny-falcon** [comment]
32bit version works. It just doesn't play well with conservative GC.

I may make 32bit version more "workable" by not storing hash-sum, and changing it to closed addressing/chained version. I think, a lot of code will be common between 32bit and 64bit versions, so they will co-exists in one file with several `if flag?(:32bit)`.

**Comment by RX14** [comment]
Perhaps seperating the "implementation-specific" and "helper-methods" parts of `hash.cr` could keep it more organised. Keep the implementation-specific methods which are changed by this PR at the bottom of the file, and everything else above the file. We can then have a big comment which explains the implementation halfway down the file before all the implementation.

**Comment by funny-falcon** [comment]
I was a bit busy with work and family. I plan to spend some time on weekends.

**Comment by funny-falcon** [comment]
damn :-( now 64bit version fails with "Repeated allocation of very large block"... I can't understand, why ;-(

**Comment by funny-falcon** [comment]
And I can't reproduce on my notebook :-( My version of libgc doesn't complain against. As well as circleci.

**Comment by akzhan** [comment]
Can you ask @ivmai?

**Comment by yxhuvud** [comment]
I thought the repeated allocation was just a warning. I get that in my implementation too.

**Comment by RX14** [comment]
No, the error is that the compiler specs ran out of memory. Check the usage before and after locally.

**Comment by funny-falcon** [comment]
Problem is I cann't reproduce. And circleci cann't as well. So, looks like it depends on hosts libgc.

I have one suggestion: probably it is because old code doesn't reallock huge bins array on `#clean` . But to check I need refactor code.

**Comment by funny-falcon** [comment]
I turned off deallocation on clear, but looks like it doesn't help.
I don't know what to do further :-(

**Comment by RX14** [comment]
@funny-falcon it's unlikely to be a libgc bug. You can't reproduce it simply because you have more ram than the travis VM.

**Comment by funny-falcon** [comment]
@RX14, I don't claim it is libgc bug. I've said, libgc behaves differently between travis-ci and my computer. Unfortunately, on my pc I have no even warnings. So it complicates investigation a lot.

**Comment by funny-falcon** [comment]
On ~5minute point (~610 samples) master consumed 490MB and patched 390MB, therefore it is not direct memory consumption, unfortunately.

**Comment by RX14** [comment]
@funny-falcon that's very strange.

**Comment by vlazar** [comment]
@funny-falcon Do you think it would make a sense to try with latest Crystal release? It's been a while, maybe the issue disappeared? :)

**Comment by funny-falcon** [comment]
Yes, I will try

**Comment by RX14** [comment]
Yeah i'm quite sad this PR stalled, `Hash` is one of the most-used datastructures in the compiler and many other crystal programs. And it's probably the datastructure with most scope for performance changes in it's implementation. I'd love to see crystal have a well-optimized `Hash`, and see what performance effects it has, especially if it has any effect on compile times.

**Comment by vlazar** [comment]
With #8017 merged in should this one be closed?

**Comment by ysbaddaden** [comment]
Yes, closing.

**Pull request closed without merging**